crates/apollo-compiler/src/validation/validation_db.rs

```diff
@@ -180,9 +185,105 @@ pub fn validate(db: &dyn ValidationDatabase) -> Vec<ApolloDiagnostic> {
     diagnostics
 }
 
+fn validate_name_uniqueness(db: &dyn ValidationDatabase) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+
+    // Different node types use different namespaces.
+    let mut fragment_scope = HashMap::<String, (FileId, ast::Name)>::new();
+    let mut operation_scope = HashMap::new();
+    let mut directive_scope = HashMap::new();
+    let mut type_scope = HashMap::new();
+
+    let all_types = db
+        .type_definition_files()
+        .into_iter()
+        .flat_map(move |file_id| {
+            db.ast(file_id)
+                .document()
+                .syntax()
+                .children()
+                .filter_map(ast::Definition::cast)
+                // Extension names are allowed to be duplicates,
+                // and schema definitions don't have names.
+                .filter(|def| {
+                    !def.is_extension_definition()
+                        && !matches!(def, ast::Definition::SchemaDefinition(_))
+                })
+                .map(move |def| (file_id, def))
+        });
```

Comment:
should we have this as its own salsa query in hir_db?

Reply:
in theeeory this should be the only place we use this approach, anything that does not need to know about duplicate names will use the existing queries.